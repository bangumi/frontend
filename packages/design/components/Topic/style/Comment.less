@import '../../../theme/base';

@highlight-border-color: #369cf8;

.bgm-comment {
  &__header {
    display: flex;
    align-items: flex-start;
    padding-top: 20px;
    padding-bottom: 12px;
    border-bottom: 1px dashed @gray-10;

    &--reply {
      margin-left: 72px;
    }

    &--collapsed {
      padding-top: 12px;
      min-height: 22px;

      .bgm-comment__content {
        margin-top: 0;
      }
    }

    &--highlighted {
      border: 2px solid @highlight-border-color;
      border-radius: 5px;
    }
  }

  &__box {
    margin-left: 12px;
    flex: 1;
    // this trick is same to EditorForm
    min-width: 0;
  }

  &__main {
    min-height: 60px;
    display: flex;
    flex-direction: column;
    align-items: flex-start;
    justify-content: space-between;
    width: 100%;
  }

  &__tip {
    font-size: 16px;
    line-height: 22px;
    display: flex;
    align-items: center;
    justify-content: space-between;
    color: @gray-60;
    width: 100%;

    .creator-info {
      display: flex;
      align-items: center;
      justify-content: space-between;

      .bgm-link,
      svg {
        padding-right: 12px;
      }
    }

    .comment-info {
      display: flex;
      font-size: 14px;
      line-height: 20px;
    }
  }

  &__opinions {
    margin-top: 12px;
  }

  &__content {
    color: @gray-100;
    line-height: 26px;
    margin-top: 12px;

    &--deleted {
      color: @gray-80;
    }
  }
}
