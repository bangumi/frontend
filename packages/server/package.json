{
  "name": "@bangumi/server",
  "version": "0.0.0",
  "private": true
}
