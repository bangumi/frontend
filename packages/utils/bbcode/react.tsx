import React from 'react';

import { convert } from './convert';
import { Parser } from './parser';
import type { CodeNodeTypes, NodeTypes, VNode } from './types';

/* eslint-disable no-useless-escape */
const toCamelCase = (s: string) => s.replace(/(\-\w)/g, (k) => k[1]?.toUpperCase() ?? '');

const joinClassName = (cls: string | string[]) => {
  if (Array.isArray(cls)) {
    return cls.join(' ');
  }
  return cls;
};

const convertToReactProps = (node: VNode, key?: React.Key) => {
  const { props, style, className } = node;

  const transformedStyle: Record<string, string> = style
    ? Object.entries(style).reduce((pre, [key, val]) => {
        return {
          ...pre,
          [toCamelCase(key)]: val,
        };
      }, {})
    : {};

  const transformedClassName = className !== undefined ? joinClassName(className) : undefined;

  const reactProps = {
    ...props,
    style: transformedStyle,
    className: transformedClassName,
    key,
  };

  return reactProps;
};

export const renderNode = (node: NodeTypes, key?: React.Key): React.ReactElement | string => {
  if (typeof node === 'string') {
    const splittedStr: Array<string | React.ReactElement> = node.split(/\n/g);

    if (splittedStr.length < 2) {
      return splittedStr[0] ?? '';
    }
    const children = splittedStr.reduce<Array<string | React.ReactElement>>((pre, curr, idx) => {
      if (idx > 0) {
        return [...pre, React.createElement('br', null), curr];
      }
      return [...pre, curr];
    }, []);
    return React.createElement(React.Fragment, { key }, ...children);
  }

  const { type, children } = node;

  if (children) {
    return React.createElement(
      type,
      convertToReactProps(node, key),
      ...children.map((child, idx) => renderNode(child, idx)),
    );
  }
  return React.createElement(type, convertToReactProps(node, key));
};

export const render = (rawStr: string): Array<React.ReactElement | string> => {
  const nodes: CodeNodeTypes[] = new Parser(rawStr).parse();

  return nodes.map((node, idx) => renderNode(convert(node, {}), idx));
};
