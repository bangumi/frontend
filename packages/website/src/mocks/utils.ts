import { rest } from 'msw'
import fsp from 'fs/promises'
import path from 'path'

async function isFileExist (filePath: string): Promise<boolean> {
  try {
    await fsp.stat(filePath)
  } catch (e) {
    return false
  }

  return true
}

async function loadFixture (pathname: string, requestMethod: string): Promise<string> {
  const fixturePath = path.join(__dirname, './fixtures', `${pathname}-GET.json`)

  if (!(await isFileExist(fixturePath))) {
    const errMessage = `缺少对应 ${pathname} API Mock 文件。请创建 ${fixturePath} 添加 JSON mock`
    console.error(errMessage)
    throw new Error(errMessage)
  }

  return JSON.parse((await fsp.readFile(fixturePath)).toString())
}

type HTTPMethods = 'get' | 'post' | 'put' | 'delete' | 'options'

export function mockAPI (url: string, method: HTTPMethods) {
  return rest[method](url, async (req, res, ctx) => {
    return res(
      ctx.status(200),
      ctx.json(await loadFixture(req.url.pathname, req.method))
    )
  })
}
