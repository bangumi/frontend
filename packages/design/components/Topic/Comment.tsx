import classNames from 'classnames';
import type { FC } from 'react';
import React, { memo, useCallback, useState } from 'react';
import { useLocation, useNavigate } from 'react-router-dom';

import { ozaClient } from '@bangumi/client';
import type { BasicReply } from '@bangumi/client/client';
import type { Reply, SubReply, User } from '@bangumi/client/topic';
import { State } from '@bangumi/client/topic';
import { Friend, OriginalPoster, TopicClosed, TopicReopen, TopicSilent } from '@bangumi/icons';
import { getUserProfileLink } from '@bangumi/utils/pages';

import Avatar from '../../components/Avatar';
import RichContent from '../../components/RichContent';
import Typography from '../../components/Typography';
import { toast } from '../Toast';
import CommentActions from './CommentActions';
import CommentInfo from './CommentInfo';
import ReplyForm from './ReplyForm';

export type CommentProps = ((SubReply & { isReply: true }) | (Reply & { isReply: false })) & {
  topicId: number;
  floor: string | number;
  originalPosterId: number;
  onCommentUpdate: () => Promise<unknown>;
  user?: User;
};

const Link = Typography.Link;

const RenderContent = memo(({ state, text }: { state: State; text: string }) => {
  switch (state) {
    case State.Normal:
      return <RichContent bbcode={text} classname='bgm-comment__content' />;
    case State.Closed:
      return <div className='bgm-comment__content'>关闭了该主题</div>;
    case State.Reopen:
      return <div className='bgm-comment__content'>重新开启了该主题</div>;
    case State.Silent:
      return <div className='bgm-comment__content'>下沉了该主题</div>;
    case State.DeletedByUser:
      return <div className='bgm-comment__content--deleted'>内容已被用户删除</div>;
    case State.DeletedByAdmin:
      return (
        <div className='bgm-comment__content--deleted'>
          内容因违反「
          <Link to='https://bgm.tv/about/guideline' isExternal>
            社区指导原则
          </Link>
          」已被删除
        </div>
      );
    default:
      return null;
  }
});

const SpecialStateIcon = memo(({ state }: { state: State }) => {
  switch (state) {
    case State.Normal:
      return null;
    case State.Closed:
      return <TopicClosed />;
    case State.Reopen:
      return <TopicReopen />;
    case State.Silent:
      return <TopicSilent />;
  }
  return null;
});

const Comment: FC<CommentProps> = ({
  text,
  creator,
  createdAt,
  floor,
  isFriend,
  originalPosterId,
  state,
  user,
  topicId,
  onCommentUpdate,
  ...props
}) => {
  const isReply = props.isReply;
  const isDeleted = state === State.DeletedByUser || state === State.DeletedByAdmin;
  // 1 关闭 2 重开 5 下沉
  const isSpecial = state === State.Closed || state === State.Reopen || state === State.Silent;
  const replies = !isReply ? props.replies : null;
  const shouldCollapse = isSpecial || (isReply && (/[+-]\d+$/.test(text) || isDeleted));
  const [collapsed, setCollapsed] = useState(shouldCollapse);

  const [showReplyEditor, setShowReplyEditor] = useState(false);
  const [replyContent, setReplyContent] = useState('');

  const elementId = `post_${props.id}`;
  const location = useLocation();
  const highlightId = location.hash.slice(1);
  const isHighlighted = highlightId === elementId;

  const headerClassName = classNames('bgm-comment__header', {
    'bgm-comment__header--reply': isReply,
    'bgm-comment__header--collapsed': collapsed,
    'bgm-comment__header--highlighted': isHighlighted,
  });

  const url = getUserProfileLink(creator.username);

  const navigate = useNavigate();

  const startReply = useCallback(() => {
    setShowReplyEditor(true);
    setReplyContent(isReply ? `[quote]${text.slice(0, 30)}[/quote]\n` : '');
  }, [isReply, text]);

  if (collapsed) {
    return (
      <div
        className={headerClassName}
        onClick={
          isSpecial
            ? undefined
            : () => {
                setCollapsed(false);
              }
        }
        id={elementId}
      >
        <span className='bgm-comment__tip'>
          <div className='creator-info'>
            <SpecialStateIcon state={state} />
            <Link to={url} isExternal>
              {creator.nickname}
            </Link>
            <RenderContent state={state} text={text} />
          </div>
          <CommentInfo createdAt={createdAt} floor={floor} isSpecial={isSpecial} />
        </span>
      </div>
    );
  }

  const handleReplySuccess = async (reply: BasicReply) => {
    // 先隐藏回复框避免scrollIntoView后布局变化
    setShowReplyEditor(false);
    navigate(`#post_${reply.id}`);
    // 刷新回复列表
    await onCommentUpdate();
  };

  const handleDeleteReply = async () => {
    if (confirm('确认删除这条回复？')) {
      const response = await ozaClient.deleteGroupPost(props.id);
      if (response.status === 200) {
        onCommentUpdate();
      } else {
        // TODO: 统一错误处理方式
        console.error(response);
        toast(response.data.message);
      }
    }
  };

  return (
    <div>
      <div className={headerClassName} id={`post_${props.id}`}>
        <Avatar
          src={isReply ? creator.avatar.medium : creator.avatar.large}
          size={isReply ? 'small' : 'medium'}
        />
        <div className='bgm-comment__box'>
          <div className='bgm-comment__main'>
            <span className='bgm-comment__tip'>
              <div className='creator-info'>
                <Link to={url} isExternal>
                  {creator.nickname}
                </Link>
                {originalPosterId === creator.id ? <OriginalPoster /> : null}
                {isFriend ? <Friend /> : null}
                {!isReply && creator.sign ? <span>{`(${creator.sign})`}</span> : null}
              </div>
              <div className='comment-info'>
                <CommentInfo createdAt={createdAt} floor={floor} id={props.id} />
                {user && !isDeleted && (
                  <>
                    &nbsp;&nbsp;|&nbsp;&nbsp;
                    <CommentActions
                      id={props.id}
                      onReply={startReply}
                      onDelete={handleDeleteReply}
                      isAuthor={user?.id === creator.id}
                      editable={!replies?.length}
                    />
                  </>
                )}
              </div>
            </span>
            <RenderContent state={state} text={text} />
          </div>
          {showReplyEditor && (
            <div className='bgm-comment__opinions'>
              <ReplyForm
                autoFocus
                topicId={topicId}
                replyTo={props.id}
                placeholder={`回复 @${creator.nickname}：`}
                content={replyContent}
                onChange={setReplyContent}
                onCancel={() => {
                  setShowReplyEditor(false);
                }}
                onSuccess={handleReplySuccess}
              />
            </div>
          )}
        </div>
      </div>
      {replies?.map((reply, idx) => (
        <Comment
          topicId={topicId}
          key={reply.id}
          isReply
          onCommentUpdate={onCommentUpdate}
          floor={`${floor}-${idx + 1}`}
          originalPosterId={originalPosterId}
          user={user}
          {...reply}
        />
      ))}
    </div>
  );
};

export default memo(Comment);
