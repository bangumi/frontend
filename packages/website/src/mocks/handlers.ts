import { mockAPI } from './utils'

const BASE_URL = 'https://api.bgm.tv'

function buildAPIURL (path: string) {
  return `${BASE_URL}${path}`
}

export const handlers = [
  mockAPI(buildAPIURL('/v0/subjects/:subjectId'), 'get'),
  mockAPI(buildAPIURL('/v0/persons/:personId/characters'), 'get'),
  mockAPI(buildAPIURL('/v0/characters/:characterId'), 'get')
]
