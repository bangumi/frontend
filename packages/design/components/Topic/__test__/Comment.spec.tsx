import { fireEvent, render as _render, waitFor } from '@testing-library/react';
import dayjs from 'dayjs';
import { rest } from 'msw';
import React from 'react';
import { MemoryRouter } from 'react-router-dom';

import { server as mockServer } from '@bangumi/website/mocks/server';

import type { CommentProps } from '../Comment';
import Comment from '../Comment';
import repliesComment from './fixtures/repliesComment.json';
import singleComment from './fixtures/singleComment.json';
import specialComment from './fixtures/specialComment.json';
import mockedCurrentUser from './fixtures/user.json';

function render(component: React.ReactElement, routerEntries?: string[]) {
  return _render(<MemoryRouter initialEntries={routerEntries}>{component}</MemoryRouter>);
}

// 0 正常评论 6 被用户删除 7 违反社区指导原则，已被删除
describe('Normal Comment', () => {
  function buildProps(
    isReply = false,
    comment?: any,
    floor = '233',
    originalPosterId = 233,
    user = mockedCurrentUser,
  ) {
    const reply = repliesComment.replies[0];
    const mockedComment = comment ?? (isReply ? reply : singleComment);

    const commentProps: CommentProps = {
      ...mockedComment,
      createdAt: dayjs(mockedComment.created_at).unix(),
      floor,
      originalPosterId,
      user,
      isReply,
    };
    return commentProps;
  }

  it.each([0, 6, 7])('should render %d', (state) => {
    const props = buildProps();
    const { container } = render(<Comment {...props} state={state} />);
    expect(container).toMatchSnapshot();
  });

  it('should render with reply', () => {
    const props = buildProps(false, repliesComment);
    const { container } = render(<Comment {...props} />);
    expect(container).toMatchSnapshot();
  });

  it('reply should have reply styles', () => {
    const props = buildProps(true);
    const { container } = render(<Comment {...props} />);
    // should have reply styles
    expect(container.getElementsByClassName('bgm-comment__header--reply').length).toBe(1);

    // should not have collapsed styles
    expect(container.getElementsByClassName('bgm-comment__header--collapsed').length).toBe(0);
  });

  it('reply end with +1/-1 reply should be collapsed', () => {
    const props = buildProps(true);
    const { container: container1 } = render(<Comment {...props} text='233+123' />);
    expect(container1.getElementsByClassName('bgm-comment__header--collapsed').length).toBe(1);

    const { container: container2 } = render(<Comment {...props} text='233-123' />);
    expect(container2.getElementsByClassName('bgm-comment__header--collapsed').length).toBe(1);

    // should not have collapsed styles if is not reply
    const { container: container3 } = render(<Comment {...buildProps(false)} text='233-123' />);
    expect(container3.getElementsByClassName('bgm-comment__header--collapsed').length).toBe(0);
  });

  it('show icons', () => {
    const props = buildProps(false);
    const { container: container1 } = render(
      <Comment {...props} isFriend originalPosterId={233} />,
    );
    expect(container1.getElementsByClassName('creator-info')[0]!.childNodes).toHaveLength(3);

    const { container: container2 } = render(
      <Comment {...props} isFriend={false} originalPosterId={1} />,
    );
    expect(container2.getElementsByClassName('creator-info')[0]!.childNodes).toHaveLength(3);

    const { container: container3 } = render(<Comment {...props} isFriend originalPosterId={1} />);
    expect(container3.getElementsByClassName('creator-info')[0]!.childNodes).toHaveLength(4);
  });

  it('show edit and delete button if current user is comment creator', () => {
    const user = { ...mockedCurrentUser, id: 1 };
    const props = buildProps(false, singleComment, '233', 233, user);
    const { getByText } = render(<Comment {...props} />);
    expect(getByText('编辑')).toBeInTheDocument();
    expect(getByText('删除')).toBeInTheDocument();
  });

  it('hide edit button if there are subreplies', () => {
    const user = { ...mockedCurrentUser, id: 1 };
    const props = buildProps(false, repliesComment, '233', 233, user);
    const { container } = render(<Comment {...props} />);
    // 选取主评论的操作区域
    const actions = container.querySelector('.bgm-comment__box .bgm-comment-actions')?.textContent;
    expect(actions?.includes('编辑')).toBeFalsy();
    expect(actions?.includes('删除')).toBeTruthy();
  });

  it('do not show opinions if not login', () => {
    const props = buildProps(false, singleComment, '233', 233, null as any);
    const { container } = render(<Comment {...props} />);
    expect(container).toMatchSnapshot();
  });

  it('click reply button should show editor form', () => {
    const props = buildProps(false);
    const { getByText, container, getByTitle } = render(<Comment {...props} />);
    expect(container.getElementsByClassName('bgm-editor__form').length).toBe(0);

    fireEvent.click(getByTitle('回复'));
    expect(container.getElementsByClassName('bgm-editor__form').length).toBe(1);

    fireEvent.click(getByText('取消'));
    expect(container.getElementsByClassName('bgm-editor__form').length).toBe(0);
  });

  it('successful reply should refresh, highlight and hide form otherwise not', async () => {
    const basicReply = { id: 2104702 };
    const mockApi = (status: number) => {
      mockServer.use(
        rest.post('/p1/groups/-/topics/1/replies', async (_, res, ctx) =>
          res(ctx.status(status), ctx.json(basicReply)),
        ),
      );
    };

    const onSuccess = vi.fn();
    const props = buildProps(false);
    const { getByText, container, getByTitle } = render(
      <Comment {...props} onCommentUpdate={onSuccess} topicId={1} />,
    );
    const fillAndSubmit = () => {
      fireEvent.click(getByTitle('回复'));
      fireEvent.change(container.querySelector('textarea')!, { target: { value: '233' } });
      fireEvent.click(getByText('写好了'));
    };

    mockApi(200);
    fillAndSubmit();
    await waitFor(() => {
      expect(onSuccess).toBeCalled();
      expect(document.getElementById(`post_${basicReply.id}`)).toHaveClass(
        'bgm-comment__header--highlighted',
      );
      expect(container.getElementsByClassName('bgm-editor__form').length).toBe(0);
    });

    onSuccess.mockClear();
    mockApi(400);
    fillAndSubmit();
    await waitFor(() => {
      expect(onSuccess).not.toBeCalled();
      expect(container.getElementsByClassName('bgm-editor__form').length).toBe(1);
    });
  });

  it('should highlight comment corresponding to hash', () => {
    const props = buildProps(false, repliesComment);
    const { container } = render(<Comment {...props} />, ['/groups/topics/1#post_2104702']);
    expect(container).toMatchSnapshot();
  });
});

// 1 关闭 2 重开 5 下沉
describe('Special Comment', () => {
  function buildProps(state: number) {
    return {
      ...specialComment,
      createdAt: dayjs(specialComment.created_at).unix(),
      state,
    } as unknown as CommentProps;
  }

  it.each([1, 2, 5])('should render state is %d', (state) => {
    const props = buildProps(state);
    const { container } = render(<Comment {...props} />);
    expect(container).toMatchSnapshot();
  });
});
