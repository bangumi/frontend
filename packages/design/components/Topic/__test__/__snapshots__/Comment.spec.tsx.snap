// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Normal Comment > do not show opinions if not login 1`] = `
<div>
  <div>
    <div
      class="bgm-comment__header"
      id="post_2104702"
    >
      <div
        class="bgm-avatar bgm-avatar--medium"
      >
        <img
          src="https://lain.bgm.tv/pic/user/l/icon.jpg"
        />
      </div>
      <div
        class="bgm-comment__box"
      >
        <div
          class="bgm-comment__main"
        >
          <span
            class="bgm-comment__tip"
          >
            <div
              class="creator-info"
            >
              <a
                class="bgm-link"
                href="https://bgm.tv/user/123456789"
              >
                用户
              </a>
              <span>
                (我是签名)
              </span>
            </div>
            <div
              class="comment-info"
            >
              <span
                class="bgm-topic__commentInfo"
              >
                <a
                  href="#post_2104702"
                >
                  #
                  233
                </a>
                  
                |
                  
                2022-9-22 06:03
              </span>
            </div>
          </span>
          <div
            class="bgm-rich-content bgm-comment__content"
          >
            123456789~~~
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Normal Comment > should highlight comment corresponding to hash 1`] = `
<div>
  <div>
    <div
      class="bgm-comment__header bgm-comment__header--highlighted"
      id="post_2104702"
    >
      <div
        class="bgm-avatar bgm-avatar--medium"
      >
        <img
          src="https://lain.bgm.tv/pic/user/l/icon.jpg"
        />
      </div>
      <div
        class="bgm-comment__box"
      >
        <div
          class="bgm-comment__main"
        >
          <span
            class="bgm-comment__tip"
          >
            <div
              class="creator-info"
            >
              <a
                class="bgm-link"
                href="https://bgm.tv/user/123456789"
              >
                用户
              </a>
              <span>
                (我是签名)
              </span>
            </div>
            <div
              class="comment-info"
            >
              <span
                class="bgm-topic__commentInfo"
              >
                <a
                  href="#post_2104702"
                >
                  #
                  233
                </a>
                  
                |
                  
                2022-9-22 06:03
              </span>
                |  
              <div
                class="bgm-comment-actions"
              >
                <button
                  class="bgm-button bgm-button--plain bgm-button--size-small"
                  title="回复"
                >
                  <div />
                </button>
                <div
                  class="bgm-popover"
                >
                  <button
                    class="bgm-button bgm-comment-actions__more bgm-button--plain bgm-button--size-small"
                    title="其他"
                  >
                    <div />
                  </button>
                  <div
                    class="bgm-popover__container"
                  >
                    <div
                      class="bgm-popover__content"
                    >
                      <div
                        class="bgm-comment-actions__popover"
                      >
                        <button
                          class="bgm-button bgm-button--text"
                        >
                          绝交
                        </button>
                        <button
                          class="bgm-button bgm-button--text"
                        >
                          报告疑虑
                        </button>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </span>
          <div
            class="bgm-rich-content bgm-comment__content"
          >
            123456789~~~
          </div>
        </div>
      </div>
    </div>
    <div>
      <div
        class="bgm-comment__header bgm-comment__header--reply bgm-comment__header--highlighted"
        id="post_2104702"
      >
        <div
          class="bgm-avatar bgm-avatar--small"
        >
          <img
            src="https://lain.bgm.tv/pic/user/m/icon.jpg"
          />
        </div>
        <div
          class="bgm-comment__box"
        >
          <div
            class="bgm-comment__main"
          >
            <span
              class="bgm-comment__tip"
            >
              <div
                class="creator-info"
              >
                <a
                  class="bgm-link"
                  href="https://bgm.tv/user/123456789"
                >
                  用户
                </a>
              </div>
              <div
                class="comment-info"
              >
                <span
                  class="bgm-topic__commentInfo"
                >
                  <a
                    href="#post_2104702"
                  >
                    #
                    233-1
                  </a>
                    
                  |
                    
                  2022-9-22 06:03
                </span>
                  |  
                <div
                  class="bgm-comment-actions"
                >
                  <button
                    class="bgm-button bgm-button--plain bgm-button--size-small"
                    title="回复"
                  >
                    <div />
                  </button>
                  <div
                    class="bgm-popover"
                  >
                    <button
                      class="bgm-button bgm-comment-actions__more bgm-button--plain bgm-button--size-small"
                      title="其他"
                    >
                      <div />
                    </button>
                    <div
                      class="bgm-popover__container"
                    >
                      <div
                        class="bgm-popover__content"
                      >
                        <div
                          class="bgm-comment-actions__popover"
                        >
                          <button
                            class="bgm-button bgm-button--text"
                          >
                            绝交
                          </button>
                          <button
                            class="bgm-button bgm-button--text"
                          >
                            报告疑虑
                          </button>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </span>
            <div
              class="bgm-rich-content bgm-comment__content"
            >
              123456789~~~
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Normal Comment > should render 0 1`] = `
<div>
  <div>
    <div
      class="bgm-comment__header"
      id="post_2104702"
    >
      <div
        class="bgm-avatar bgm-avatar--medium"
      >
        <img
          src="https://lain.bgm.tv/pic/user/l/icon.jpg"
        />
      </div>
      <div
        class="bgm-comment__box"
      >
        <div
          class="bgm-comment__main"
        >
          <span
            class="bgm-comment__tip"
          >
            <div
              class="creator-info"
            >
              <a
                class="bgm-link"
                href="https://bgm.tv/user/123456789"
              >
                用户
              </a>
              <span>
                (我是签名)
              </span>
            </div>
            <div
              class="comment-info"
            >
              <span
                class="bgm-topic__commentInfo"
              >
                <a
                  href="#post_2104702"
                >
                  #
                  233
                </a>
                  
                |
                  
                2022-9-22 06:03
              </span>
                |  
              <div
                class="bgm-comment-actions"
              >
                <button
                  class="bgm-button bgm-button--plain bgm-button--size-small"
                  title="回复"
                >
                  <div />
                </button>
                <div
                  class="bgm-popover"
                >
                  <button
                    class="bgm-button bgm-comment-actions__more bgm-button--plain bgm-button--size-small"
                    title="其他"
                  >
                    <div />
                  </button>
                  <div
                    class="bgm-popover__container"
                  >
                    <div
                      class="bgm-popover__content"
                    >
                      <div
                        class="bgm-comment-actions__popover"
                      >
                        <button
                          class="bgm-button bgm-button--text"
                        >
                          绝交
                        </button>
                        <button
                          class="bgm-button bgm-button--text"
                        >
                          报告疑虑
                        </button>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </span>
          <div
            class="bgm-rich-content bgm-comment__content"
          >
            123456789~~~
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Normal Comment > should render 6 1`] = `
<div>
  <div>
    <div
      class="bgm-comment__header"
      id="post_2104702"
    >
      <div
        class="bgm-avatar bgm-avatar--medium"
      >
        <img
          src="https://lain.bgm.tv/pic/user/l/icon.jpg"
        />
      </div>
      <div
        class="bgm-comment__box"
      >
        <div
          class="bgm-comment__main"
        >
          <span
            class="bgm-comment__tip"
          >
            <div
              class="creator-info"
            >
              <a
                class="bgm-link"
                href="https://bgm.tv/user/123456789"
              >
                用户
              </a>
              <span>
                (我是签名)
              </span>
            </div>
            <div
              class="comment-info"
            >
              <span
                class="bgm-topic__commentInfo"
              >
                <a
                  href="#post_2104702"
                >
                  #
                  233
                </a>
                  
                |
                  
                2022-9-22 06:03
              </span>
            </div>
          </span>
          <div
            class="bgm-comment__content--deleted"
          >
            内容已被用户删除
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Normal Comment > should render 7 1`] = `
<div>
  <div>
    <div
      class="bgm-comment__header"
      id="post_2104702"
    >
      <div
        class="bgm-avatar bgm-avatar--medium"
      >
        <img
          src="https://lain.bgm.tv/pic/user/l/icon.jpg"
        />
      </div>
      <div
        class="bgm-comment__box"
      >
        <div
          class="bgm-comment__main"
        >
          <span
            class="bgm-comment__tip"
          >
            <div
              class="creator-info"
            >
              <a
                class="bgm-link"
                href="https://bgm.tv/user/123456789"
              >
                用户
              </a>
              <span>
                (我是签名)
              </span>
            </div>
            <div
              class="comment-info"
            >
              <span
                class="bgm-topic__commentInfo"
              >
                <a
                  href="#post_2104702"
                >
                  #
                  233
                </a>
                  
                |
                  
                2022-9-22 06:03
              </span>
            </div>
          </span>
          <div
            class="bgm-comment__content--deleted"
          >
            内容因违反「
            <a
              class="bgm-link"
              href="https://bgm.tv/about/guideline"
            >
              社区指导原则
            </a>
            」已被删除
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Normal Comment > should render with reply 1`] = `
<div>
  <div>
    <div
      class="bgm-comment__header"
      id="post_2104702"
    >
      <div
        class="bgm-avatar bgm-avatar--medium"
      >
        <img
          src="https://lain.bgm.tv/pic/user/l/icon.jpg"
        />
      </div>
      <div
        class="bgm-comment__box"
      >
        <div
          class="bgm-comment__main"
        >
          <span
            class="bgm-comment__tip"
          >
            <div
              class="creator-info"
            >
              <a
                class="bgm-link"
                href="https://bgm.tv/user/123456789"
              >
                用户
              </a>
              <span>
                (我是签名)
              </span>
            </div>
            <div
              class="comment-info"
            >
              <span
                class="bgm-topic__commentInfo"
              >
                <a
                  href="#post_2104702"
                >
                  #
                  233
                </a>
                  
                |
                  
                2022-9-22 06:03
              </span>
                |  
              <div
                class="bgm-comment-actions"
              >
                <button
                  class="bgm-button bgm-button--plain bgm-button--size-small"
                  title="回复"
                >
                  <div />
                </button>
                <div
                  class="bgm-popover"
                >
                  <button
                    class="bgm-button bgm-comment-actions__more bgm-button--plain bgm-button--size-small"
                    title="其他"
                  >
                    <div />
                  </button>
                  <div
                    class="bgm-popover__container"
                  >
                    <div
                      class="bgm-popover__content"
                    >
                      <div
                        class="bgm-comment-actions__popover"
                      >
                        <button
                          class="bgm-button bgm-button--text"
                        >
                          绝交
                        </button>
                        <button
                          class="bgm-button bgm-button--text"
                        >
                          报告疑虑
                        </button>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </span>
          <div
            class="bgm-rich-content bgm-comment__content"
          >
            123456789~~~
          </div>
        </div>
      </div>
    </div>
    <div>
      <div
        class="bgm-comment__header bgm-comment__header--reply"
        id="post_2104702"
      >
        <div
          class="bgm-avatar bgm-avatar--small"
        >
          <img
            src="https://lain.bgm.tv/pic/user/m/icon.jpg"
          />
        </div>
        <div
          class="bgm-comment__box"
        >
          <div
            class="bgm-comment__main"
          >
            <span
              class="bgm-comment__tip"
            >
              <div
                class="creator-info"
              >
                <a
                  class="bgm-link"
                  href="https://bgm.tv/user/123456789"
                >
                  用户
                </a>
              </div>
              <div
                class="comment-info"
              >
                <span
                  class="bgm-topic__commentInfo"
                >
                  <a
                    href="#post_2104702"
                  >
                    #
                    233-1
                  </a>
                    
                  |
                    
                  2022-9-22 06:03
                </span>
                  |  
                <div
                  class="bgm-comment-actions"
                >
                  <button
                    class="bgm-button bgm-button--plain bgm-button--size-small"
                    title="回复"
                  >
                    <div />
                  </button>
                  <div
                    class="bgm-popover"
                  >
                    <button
                      class="bgm-button bgm-comment-actions__more bgm-button--plain bgm-button--size-small"
                      title="其他"
                    >
                      <div />
                    </button>
                    <div
                      class="bgm-popover__container"
                    >
                      <div
                        class="bgm-popover__content"
                      >
                        <div
                          class="bgm-comment-actions__popover"
                        >
                          <button
                            class="bgm-button bgm-button--text"
                          >
                            绝交
                          </button>
                          <button
                            class="bgm-button bgm-button--text"
                          >
                            报告疑虑
                          </button>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </span>
            <div
              class="bgm-rich-content bgm-comment__content"
            >
              123456789~~~
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Special Comment > should render state is 1 1`] = `
<div>
  <div
    class="bgm-comment__header bgm-comment__header--collapsed"
    id="post_2056216"
  >
    <span
      class="bgm-comment__tip"
    >
      <div
        class="creator-info"
      >
        <div />
        <a
          class="bgm-link"
          href="https://bgm.tv/user/123456789"
        >
          用户
        </a>
        <div
          class="bgm-comment__content"
        >
          关闭了该主题
        </div>
      </div>
      <span
        class="bgm-topic__commentInfo"
      >
        2022-7-17 04:08
      </span>
    </span>
  </div>
</div>
`;

exports[`Special Comment > should render state is 2 1`] = `
<div>
  <div
    class="bgm-comment__header bgm-comment__header--collapsed"
    id="post_2056216"
  >
    <span
      class="bgm-comment__tip"
    >
      <div
        class="creator-info"
      >
        <div />
        <a
          class="bgm-link"
          href="https://bgm.tv/user/123456789"
        >
          用户
        </a>
        <div
          class="bgm-comment__content"
        >
          重新开启了该主题
        </div>
      </div>
      <span
        class="bgm-topic__commentInfo"
      >
        2022-7-17 04:08
      </span>
    </span>
  </div>
</div>
`;

exports[`Special Comment > should render state is 5 1`] = `
<div>
  <div
    class="bgm-comment__header bgm-comment__header--collapsed"
    id="post_2056216"
  >
    <span
      class="bgm-comment__tip"
    >
      <div
        class="creator-info"
      >
        <div />
        <a
          class="bgm-link"
          href="https://bgm.tv/user/123456789"
        >
          用户
        </a>
        <div
          class="bgm-comment__content"
        >
          下沉了该主题
        </div>
      </div>
      <span
        class="bgm-topic__commentInfo"
      >
        2022-7-17 04:08
      </span>
    </span>
  </div>
</div>
`;
